File: day_01/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <numeric>

using ints = std::vector<int>;

std::pair<ints, ints> ParseInput(const std::string& fname)
{
    const auto lines = pypp::splitFileLines(fname);

    ints first_ids {};
    ints second_ids {};
    first_ids.reserve(lines.size());
    second_ids.reserve(lines.size());

    for (const auto& line : lines) {
        first_ids.push_back(std::stoi(pypp::splitThenGetAt(line, ' ', 0)));
        second_ids.push_back(std::stoi(pypp::splitThenGetAt(line, ' ', 1)));
    }

    return { first_ids, second_ids };
}

ints ComputeDistances(ints list_a, ints list_b)
{
    ints result {};
    std::ranges::sort(list_a);
    std::ranges::sort(list_b);

    for (int i { 0 }; i < list_a.size(); ++i)
        result.push_back(std::abs(list_a.at(i) - list_b.at(i)));

    return result;
}

int main()
{
    auto [ids_1, ids_2] = ParseInput("input.txt");

    const auto distances { ComputeDistances(ids_1, ids_2) };
    std::cout << std::accumulate(std::begin(distances), std::end(distances), 0) << std::endl;

    return 0;
}

File: day_01/p1.py
def parse_input(fname: str):
    lines = []
    with open(fname, 'r') as f:
        lines = f.read().splitlines()
    
    pairs = [map(int, line.split()) for line in lines]
    return list(zip(*pairs))

def compute_distance(a, b):
    a = sorted(a)
    b = sorted(b)

    return [abs(elem[0]-elem[1]) for elem in zip(a,b)]
    
def main():
    input = parse_input("input.txt")
    first = input[0]
    second = input[1]
    
    print(sum(compute_distance(first, second)))

if __name__ == "__main__":
    main()
